### ISSUE TITLE
build.d foregoes catching advantage?

### ISSUE THREAD

**Issue by blaggacao** [Tecnativa/docker-odoo-base#24]
Hi Jairo, I think using build.d technique forgoes the catching advantage of individual `COPY file RUN file` combinations, as the cache on copy validates the file hash. Something that docker cannot infer on your build.d script.  For time intensive build instructions, this can be quite some annoyance. Or am I missing something?

**Comment by blaggacao** [comment]
I'm therefore in the process of adapting the folder structure to something like this:
(Opinionating for a better control of the building blocks not through "magic" but through descriptive dockerfiles. 😉 )

```markdown
Basically, every directory you have to worry about is found inside `/opt/odoo`.
This is its structure:

    addons/
        ...
        80-odoo-ee/
        90-odoo-cc/
    framework/
        addons/
        odoo/
        odoo-bin
        requirements.txt
    odoo.conf

Docker boilerplate is living in `/home`. It mainly constructs the config file.

    common/
        entrypoint.d/
        conf.d/
    custom/
        entrypoint.d/
        conf.d/

PATH and PYTHONPATH enabled scripts to consider. For convenience.

    ./bin/*    -> /usr/local/bin/
    ./lib/*.py -> /usr/local/lib/python2.7/dist-packages/
```

**Comment by blaggacao** [comment]
Idea, something like this:
```Dockerfile
# --- User Creation ---
ADD run/create-user.sh /tmp/command
RUN chmod +x /tmp/command && sync && /tmp/command

# --- Dependency Instalation ---
ADD run/apt-get.sh /tmp/command
RUN chmod +x /tmp/command && sync && /tmp/command

ADD run/pip-npm-installs.sh /tmp/command
ADD run/requirements.txt /tmp/
RUN chmod +x /tmp/command && sync && /tmp/command

ADD run/wkhtmltopdf-install-0.12.3.sh /tmp/command
RUN chmod +x /tmp/command && sync && /tmp/command
```

**Comment by Yajo** [comment]
I guess you're talking aboug *caching*, not *catching*...

You can create a script like this in your [`custom/build.d` dir](https://github.com/Tecnativa/docker-odoo-base/tree/scaffolding/odoo/custom/build.d):

```bash
#!/bin/bash
mkdir -p /opt/odoo/common/build-step-by-step
mv /opt/odoo/common/build.d/* /opt/odoo/common/build-step-by-step/
```

Then that would invalidate the single-step cache in your project. Then you can add in your project's `Dockerfile`:

```Dockerfile
FROM tecnativa/odoo-base:10.0
RUN /opt/odoo/common/build-step-by-step/20-repos-aggregate
RUN /opt/odoo/common/build-step-by-step/30-addons-link
RUN /opt/odoo/common/build-step-by-step/40-clean
RUN /opt/odoo/common/build-step-by-step/50-compile
RUN /opt/odoo/common/build-step-by-step/60-config-generate
RUN /opt/odoo/common/build-step-by-step/70-odoo-install
RUN /opt/odoo/common/build-step-by-step/80-requirements-txt
```

To get specific caching per step.

However this is definitely not something we want to do in each of our projects. Disables many automatisms that make this image useful and projects based on it maintainable.

I guess we'd consider accepting a PR that moves all of those commands to `ONBUILD` directives (instead of having just one `/opt/odoo/common/build.sh` step as we have now), but not changing the structure layout, which is pretty solid right now (and I see no added value by moving things to `/home`, honestly 🤔 ).

Closing, since there's nothing left to do on our side AFAICS. Feel free to reopen otherwise. 😊

**Comment by blaggacao** [comment]
Thanks, for your insightful opinion! 
My intent, btw, is not changing the structure layout. It justs fits better our current and future repo structure, which has not that "project focus", this base image was build in mind with.

I wonder however about the correct cache invalidation of your suggestion (I my reasoning is right):

- Adding or changing a file in build.d retriggers an initial `copy` directive, gladfully `onbuild` so as to not alter the original base setup
-  As a first file, we add the suggested mv script in order to empty this recently copied
- We execute the scripts step by step.

If I'm not wrong here in something, this means that on any change or addition we will trigger the initial copy directive, effectively sabotaging the later attempt to use caching.

I guess `ADD RUN` - combos are still the most reliable and transparent way of doing it. 
Of course the inevitable tradeoff is losing magic while gaining control. But that's something that might pay off on the build pipeline of a still rather unstable project with respect to its dependencies.

**Comment by blaggacao** [comment]
My final layout looks like this:
```
    addons/
        ...
        80-odoo-ee/
        90-odoo-cc/
    odoo/
        ...
    odoo-bin
    odoo.conf
```
It "forces" you to clearly specify you mounting points when working in local development by separating framework and community addons.

**Comment by blaggacao** [comment]
@Yajo https://github.com/sameersbn/docker-apt-cacher-ng#usage Interesting! A local apt caching service, especially for if you are on slow inet connection...

**Comment by lasley** [comment]
Also good for when your servers don't have a direct route to the internet

**Comment by blaggacao** [comment]
[squid-in-a-can](https://github.com/jpetazzo/squid-in-a-can) is even better, proxies every non-encrypted traffic, so it also works great on non-https curls...

**Comment by Yajo** [comment]
Hi there, I don't understand a thing 😅. Not sure what's exactly your proposal/enhancement/bug/problem, but let me explain my points and red lines:

1. apt cache != Docker cache. I was talking about Docker cache all the time, were you speaking about apt cache instead?
2. Entrypoints affect downtime, build scripts do not. Shorter downtime is more important than shorter build time, so if something can exist in build time, it will not be moved to an entrypoint.
3. `addons.yaml` file is not only used to enable or disable certain addons, but also to actually remove their source code if they are not used, or link them where they are expected to be. That happens at build time too.
1. The usefulness of having a `custom/build.d` directory is that otherwise you wouldn't be able to run anything before the `ONBUILD` sentences. If you want to run them after, simply add `RUN`s to your Dockerfile.
3. Project structure will stay as it is, unless there's a proven benefit on changing it.

   This project aims to provide a standard layout for a Docker-based PaaS Odoo deployment, solving any problems found in the past based on our experience. In the road of defining a *standard*, you have to make decisions. Our decision was to make the directory structure hardcoded, similar to how an Ansible playbook expects some files to be in some folders, or similar to how `__main__.py` and `__init__.py` files have a special meaning for Python if found in a folder.

   What would Guido tell you if you said to him that you want your Python modules to load `__start__.py` instead of `__main__.py`? I guess the same as me here... 😁 
4. This is on production today.
4. `PATH` already includes `/usr/local/bin`, put your scripts there.
3. `PYTHONPATH` is unneeded to be patched. Just install through apt/pip and it will be in the `PYTHONPATH`.

> It "forces" you to clearly specify your mounting points when working in local development by separating framework and community addons.

And, another thing... Have you actually tried this? 🤔  We provide a scaffolding that already tells you where to put framework and community addons...

BTW, I guess that if we are talking about some code changes, better keep on speaking over a PR than over an issue, because code speaks better than words IMHO; I'd understand all better with code.

Thanks!

**Comment by blaggacao** [comment]
Hi Jairo, I'll answer inline 😄 Thanks for pushing the discussion!

>apt cache != Docker cache. I was talking about Docker cache all the time, were you speaking about apt cache instead?

I was talking about docker cache, but apt cache or even better squid cache can remedy some shortcomings, where docker cache doesn't catch up. 

>Entrypoints affect downtime, build scripts do not. Shorter downtime is more important than shorter build time, so if something can exist in build time, it will not be moved to an entrypoint.

I guess, that's only true, if you assume not doing rolling updates. I cannot see how a running container, which after startup signals to the helcheck service "I'm up" would affect downtime.

>addons.yaml file is not only used to enable or disable certain addons, but also to actually remove their source code if they are not used, or link them where they are expected to be. That happens at build time too.

I admire the sophistication and I guess this is the perfect tool for the OCA multiverse. Another flavor of achieving similar things is git submodule or git subtree. addons.yml has a really nice mix-and-match aspect, which I was long looking for. However, personally, we do more monolithic development at the gain of efficiency and control, but on the loss of flexibility.

>The usefulness of having a custom/build.d directory is that otherwise you wouldn't be able to run anything before the ONBUILD sentences. If you want to run them after, simply add RUNs to your Dockerfile.

I didn't want to forgo the build cache, so what I did is downstep the base builds and made individual builds, while maintaining the build.d logic on the ONBUILD. This way it brings automation to the relatively small date of project dockers, while maintaining control and speed on base image refactoring. It suits our use case better. (I'll release it as soon as it is accepted in our upstream repo and I move on to our factoring it into a separate repo)

> Project structure will stay as it is, unless there's a proven benefit on changing it.

I didn't want to challenge that, at all, actually. It is very well designed for the (implicitly) proposed use case, I guess. In my case, I have: 1. Odoo Enterprise 2. DevCO-Felicity 2a. DevCO-Felicity-MgtPlatform 2b. DevCO-Felicity-Textile 2c. DevCO-Felicity-Construccion etcetera... It's clear downstream hirachy, less mix and match. Again, I'll publish my solution soon, so you might have a look.

>This is on production today.

That's a fortunate fact, I guess 😉 

> PATH already includes /usr/local/bin, put your scripts there.

Yep, I saw that on a forth read. I pretty much like the pattern, because it get's folders in a pretty much single purpose docker out of the way. "Folder" (ie structure) and "single purpose" are kind of two of a kind...

>PYTHONPATH is unneeded to be patched. Just install through apt/pip and it will be in the PYTHONPATH.

That's also a fortunate fact, I guess, which enables the above pattern, just for python scripts.

>>It "forces" you to clearly specify your mounting points when working in local development by separating framework and community addons.

>And, another thing... Have you actually tried this? 🤔 We provide a scaffolding that already tells you where to put framework and community addons...

I could not make real fit with the scaffolding, especially, as I was adapting our legacy structure, which is in production right now, augmenting it with some of your's and camptocamp's ideas (see marvelous [marabunta] (https://github.com/camptocamp/marabunta)).

Ok I'll just copy it and publish now... 😄
